erdjs/src/smartcontracts/codec/binary.ts

```diff
@@ -0,0 +1,119 @@
+import * as errors from "../../errors";
+import { FunctionDefinition, onTypedValueSelect, onTypeSelect, PrimitiveType, StructureType, TypeDescriptor, TypedValue, U8Type } from "../typesystem";
+import { guardSameLength } from "../../utils";
+import { OptionalValueBinaryCodec } from "./optional";
+import { PrimitiveBinaryCodec } from "./primitive";
+import { VectorBinaryCodec } from "./vector";
+import { StructureBinaryCodec } from "./structure";
+
+export class BinaryCodec {
+    readonly constraints: BinaryCodecConstraints;
+    private readonly optionalCodec: OptionalValueBinaryCodec;
+    private readonly vectorCodec: VectorBinaryCodec;
+    private readonly primitiveCodec: PrimitiveBinaryCodec;
+    private readonly structureCodec: StructureBinaryCodec;
+    
+    constructor(constraints: BinaryCodecConstraints | null = null) {
+        this.constraints = constraints || new BinaryCodecConstraints();
+        this.optionalCodec = new OptionalValueBinaryCodec(this);
+        this.vectorCodec = new VectorBinaryCodec(this);
+        this.primitiveCodec = new PrimitiveBinaryCodec(this);
+        this.structureCodec = new  StructureBinaryCodec(this);
+    }
+
+    decodeFunctionOutput(outputItems: Buffer[], definition: FunctionDefinition): TypedValue[] {
+        guardSameLength(outputItems, definition.output);
+
+        let result: TypedValue[] = [];
+
+        // For output parameters, top-level decoding is normally used.
+        // TODO: Question for review - is this assumption correct?
+        for (let i = 0; i < outputItems.length; i++) {
```

Comment:
Yes, each output is encoded as top-level.

Reply:
And input arguments as well (did that in `Argument.typed()` utility) - right?

Reply:
Each argument is interpreted as top-level encoding of some type. Each top-encoded type corresponds to 1 argument, or 1 result.

If you look at the Rust source code, you will also discover some types that do not implement TopEncode/TopDecode, but can be used as argument or result types. These include: OptionalArg, OptionalResult, VarArgs, MultiResultVec, etc. These do not correspond to 1 argument/result, but to either several, or variable number of arguments/results. Their contents, however, need to implement TopEncode/TopDecode.

Also note that they don't all need to implement NestedEncode/NestedDecode to be used as arguments/results. In the contract there are some examples for types where I did not bother to implement the nested counterpart, since they never get embedded.